File: src/data/courses/dsa/binary-search-trees/bst-core-operations/kth-smallest-largest/code/java/KthSmallestLargest.java
// ============================================================
// Kth Smallest/Largest
// ============================================================
class Solution {
    int count, result;
    public int kthSmallest(TreeNode root, int k) {
        count = k; result = -1; inorder(root); return result;
    }
    void inorder(TreeNode n) {
        if (n == null || count <= 0) return;
        inorder(n.left);
        if (--count == 0) { result = n.val; return; }
        inorder(n.right);
    }
    public int kthLargest(TreeNode root, int k) {
        count = k; result = -1; revInorder(root); return result;
    }
    void revInorder(TreeNode n) {
        if (n == null || count <= 0) return;
        revInorder(n.right);
        if (--count == 0) { result = n.val; return; }
        revInorder(n.left);
    }
}

File: src/data/courses/dsa/binary-search-trees/bst-core-operations/kth-smallest-largest/code/python/kth_smallest_largest.py
# ============================================================
# Kth Smallest/Largest
# ============================================================
class Solution:
    def kthSmallest(self, root, k):
        self.k, self.result = k, -1
        def inorder(node):
            if not node or self.k <= 0: return
            inorder(node.left)
            self.k -= 1
            if self.k == 0: self.result = node.val; return
            inorder(node.right)
        inorder(root); return self.result

    def kthLargest(self, root, k):
        self.k, self.result = k, -1
        def rev_inorder(node):
            if not node or self.k <= 0: return
            rev_inorder(node.right)
            self.k -= 1
            if self.k == 0: self.result = node.val; return
            rev_inorder(node.left)
        rev_inorder(root); return self.result
